driver-core/src/main/java/com/datastax/driver/core/QueryLogger.java

```diff
@@ -0,0 +1,523 @@
+                if(statement instanceof BoundStatement) {
+                    appendParameters((BoundStatement)statement, params, maxLoggedParameters);
+                } else if (statement instanceof BatchStatement) {
+                    BatchStatement batchStatement = (BatchStatement)statement;
+                    int remaining = maxLoggedParameters;
+                    for (Statement inner : batchStatement.getStatements()) {
+                        if(inner instanceof BoundStatement) {
+                            remaining = appendParameters((BoundStatement)inner, params, remaining);
+                        }
+                    }
+                }
+                if(params.length() > 0) params.append("]");
+                logger.trace(message + params, exception);
+            } else {
+                logger.debug(message, exception);
+            }
+        }
+    }
+
+    private String statementAsString(Statement statement) {
+        StringBuilder sb = new StringBuilder();
+        append(statement, sb, maxQueryStringLength);
+        return sb.toString();
+    }
+
+    private int appendParameters(BoundStatement statement, StringBuilder buffer, int remaining) {
+        if(remaining == 0) return 0;
+        ColumnDefinitions metadata = statement.preparedStatement().getVariables();
+        int numberOfParameters = metadata.size();
+        if (numberOfParameters > 0) {
+            List<ColumnDefinitions.Definition> definitions = metadata.asList();
+            int numberOfLoggedParameters;
+            if(remaining == -1){
+                numberOfLoggedParameters = numberOfParameters;
+            } else {
+                numberOfLoggedParameters = remaining > numberOfParameters ? numberOfParameters : remaining;
+                remaining -= numberOfLoggedParameters;
+            }
+            for (int i = 0; i < numberOfLoggedParameters; i++) {
+                if(buffer.length() == 0) buffer.append(" [");
+                else buffer.append(", ");
+                buffer.append(String.format("%s:%s", metadata.getName(i), parameterValueAsString(definitions.get(i), statement.values[i])));
+            }
+            if(numberOfLoggedParameters < numberOfParameters) {
+                buffer.append(FURTHER_PARAMS_OMITTED);
+            }
+        }
+        return remaining;
+    }
+
+    private String parameterValueAsString(ColumnDefinitions.Definition definition, ByteBuffer raw) {
+        String valueStr;
+        if (raw == null || raw.remaining() == 0) {
+            valueStr = "NULL";
+        } else {
+            DataType type = definition.getType();
+            Object value = type.deserialize(raw);
+            valueStr = type.format(value);
+            int maxParameterValueLength = this.maxParameterValueLength;
+            if (maxParameterValueLength != -1 && valueStr.length() > maxParameterValueLength) {
+                valueStr = valueStr.substring(0, maxParameterValueLength) + TRUNCATED_OUTPUT;
```

Comment:
Maybe we should be smarter than this for BLOBs. If we get a 100K blob, we're going to convert it to a string at line 467, then find out that this string is too long and truncate it.
Instead, we should check the ByteBuffer's capacity first and, if it's too long, `slice()` the part we're going to print and pass that to `format`. The size of the string is `2 + 2 * sizeOfBuffer` (the initial 2 is for the "0x" prefix).
This is less of an issue for other datatypes.